noir-projects/aztec-nr/aztec/src/test/helpers/context_builder.nr

```diff
@@ -0,0 +1,52 @@
+use crate::context::{PrivateContext, PublicContext};
+use dep::protocol_types::address::AztecAddress;
+use dep::std::test::OracleMock;
+
+struct ContextBuilder {
```

Comment:
Something I don't like about the idea of a context builder is that
1. if you can produce both private and public it assumes a certain similarity, which I thought we didn't want to (I'll be removing all context traits at some point in a cleanup)
2. it modifies parts of the context that I would think are "private"

Reply:
1. I considered making the builder generic over a type `T` that is then inferred to be private or public depending on which build method you call, and then make each impl support things that are specific to each other, but this ultimately was overkill at this stage. Note that so far we only have contract address and block number, which are very basic concepts - the priv/pub builder may diverte in the future, but I don't think that'd happen soon.

2. What do you mean 'private'? Keep in mind that this is simply a mocked environment for unit testing, not something that's used in any kind of real application.

Reply:
> What do you mean 'private'?

For example, the `PrivateContext` is constructed with `new(inputs: PrivateContextInputs,...)` and then (ideally) accessed through the getters/setters. I know there's no concept of private or public, but it's clear that if you modify it's state in general you cannot assume you keep the integrity of the type.

So, creating an empty type and modifying its "internal state" is not great. I know this is only for testing, but when you had done it for the old `PublicContext`, just for block_number, it already gave problems because the new implementation does not have the same internal state. On top of all that, you are (were) implicitly assuming that modifying `.block_number` has an effect on `block_number()`. Which was true, but then it wasn't. It leaks internal knowledge of how `*Context` works. Just as mocking the oracle is not great.

IDK what are the techniques for testing/mocking in Rust/Noir. I guess this could be solved by always using traits like PublicContextInterface, and then you can use the real one or a fake one conforming to the trait in the test. But that's a massive change and PITA for not much, so we don't wan't it.

In short I think using an empty context and filling it up breaks encapsulation, and it's a brittle solution.

>Keep in mind that this is simply a mocked environment for unit testing, not something that's used in any kind of real application.

Yes but it makes assumptions over the real objects. It actually _uses_ the real objects. You had to add an `::empty()` method back then. I also had to find a way to satisfy its usage when I switched the PublicContext, so it (could have) imposed restrictions on the real thing, or I could've had to drop all tests using it.